### PR TITLE
sql: fix lease manager lock ordering

### LINKED ISSUE #12903: github.com/cockroachdb/cockroach/pkg/sql: TestLeaseManagerDrain failed under stress
SHA: https://github.com/cockroachdb/cockroach/commits/6dbac80561b565d2f5e8304b9f36c4ed9112f0f2

Parameters:
```
COCKROACH_PROPOSER_EVALUATED_KV=true
TAGS=deadlock
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=117241&tab=buildLog

```
W170113 09:54:22.304804 2279390 server/status/runtime.go:116  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I170113 09:54:22.306996 2279390 server/config.go:456  1 storage engine initialized
I170113 09:54:22.308869 2279390 server/node.go:426  [n?] store [n0,s0] not bootstrapped
I170113 09:54:22.316763 2279390 server/node.go:355  [n?] **** cluster 2f84a33a-c847-4a7e-96da-46dc90a686bb has been created
I170113 09:54:22.316848 2279390 server/node.go:356  [n?] **** add additional nodes by specifying --join=127.0.0.1:35269
I170113 09:54:22.319644 2279390 storage/store.go:1250  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170113 09:54:22.319805 2279390 server/node.go:439  [n1] initialized store [n1,s1]: {Capacity:536870912 Available:536870912 RangeCount:1 LeaseCount:0}
I170113 09:54:22.319927 2279390 server/node.go:324  [n1] node ID 1 initialized
I170113 09:54:22.320024 2279390 gossip/gossip.go:292  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:35269" > attrs:<> locality:<> 
I170113 09:54:22.320262 2279390 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I170113 09:54:22.320366 2279390 server/node.go:569  [n1] connecting to gossip network to verify cluster ID...
I170113 09:54:22.321160 2279390 server/node.go:589  [n1] node connected via gossip and verified as part of cluster "2f84a33a-c847-4a7e-96da-46dc90a686bb"
I170113 09:54:22.321232 2279390 server/node.go:374  [n1] node=1: started with [[]=] engine(s) and attributes []
I170113 09:54:22.321300 2279390 sql/executor.go:322  [n1] creating distSQLPlanner with address {tcp 127.0.0.1:35269}
I170113 09:54:22.323056 2279390 server/server.go:621  [n1] starting https server at 127.0.0.1:39457
I170113 09:54:22.323111 2279390 server/server.go:622  [n1] starting grpc/postgres server at 127.0.0.1:35269
I170113 09:54:22.323162 2279390 server/server.go:623  [n1] advertising CockroachDB node at 127.0.0.1:35269
I170113 09:54:22.328352 2285294 storage/split_queue.go:99  [split,n1,s1,r1/1:/M{in-ax},@c43d9d6f00] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I170113 09:54:22.377229 2285325 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:35269} Attrs: Locality:} ClusterID:2f84a33a-c847-4a7e-96da-46dc90a686bb StartedAt:1484301262321203479}
I170113 09:54:22.387503 2285294 storage/replica_command.go:2354  [split,n1,s1,r1/1:/M{in-ax},@c43d9d6f00] initiating a split of this range at key /Table/11 [r2]
E170113 09:54:22.415251 2285294 storage/queue.go:599  [split,n1,s1,r1/1:/{Min-Table/11},@c43d9d6f00] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
I170113 09:54:22.416838 2285294 storage/split_queue.go:99  [split,n1,s1,r2/1:/{Table/11-Max},@c438fc6000] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
I170113 09:54:22.416970 2285294 storage/replica_command.go:2354  [split,n1,s1,r2/1:/{Table/11-Max},@c438fc6000] initiating a split of this range at key /Table/12 [r3]
I170113 09:54:22.429159 2279390 sql/event_log.go:95  [n1] Event: "alter_table", target: 12, info: {TableName:eventlog Statement:ALTER TABLE system.eventlog ALTER COLUMN uniqueID SET DEFAULT uuid_v4() User:node MutationID:0 CascadeDroppedViews:[]}
E170113 09:54:22.440887 2285294 storage/queue.go:599  [split,n1,s1,r2/1:/Table/1{1-2},@c438fc6000] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
I170113 09:54:22.442750 2285294 storage/split_queue.go:99  [split,n1,s1,r3/1:/{Table/12-Max},@c43c2c4900] splitting at keys [/Table/13/0 /Table/14/0]
I170113 09:54:22.443011 2285294 storage/replica_command.go:2354  [split,n1,s1,r3/1:/{Table/12-Max},@c43c2c4900] initiating a split of this range at key /Table/13 [r4]
E170113 09:54:22.503388 2285294 storage/queue.go:599  [split,n1,s1,r3/1:/Table/1{2-3},@c43c2c4900] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
I170113 09:54:22.503979 2285294 storage/split_queue.go:99  [split,n1,s1,r4/1:/{Table/13-Max},@c43c01b200] splitting at keys [/Table/14/0]
I170113 09:54:22.504079 2285294 storage/replica_command.go:2354  [split,n1,s1,r4/1:/{Table/13-Max},@c43c01b200] initiating a split of this range at key /Table/14 [r5]
I170113 09:54:22.535958 2279390 server/server.go:676  [n1] done ensuring all necessary migrations have run
I170113 09:54:22.536044 2279390 server/server.go:678  [n1] serving sql connections
POTENTIAL DEADLOCK: Inconsistent locking. saw this ordering in one goroutine:
happened before
lease.go:714 sql.(*tableState).release { t.mu.Lock() } <<<<<
lease.go:1153 sql.(*LeaseManager).Release { return t.release(lease, m) }
lease_test.go:133 sql_test.(*leaseTest).release { return t.node(nodeID).Release(lease) }
lease_test.go:147 sql_test.(*leaseTest).mustRelease { if err := t.release(nodeID, lease); err != nil { }
lease_test.go:427 sql_test.TestLeaseManagerDrain { t.mustRelease(1, l1, nil) }

happened after
lease.go:1157 sql.(*LeaseManager).isDraining { m.mu.Lock() } <<<<<
lease.go:732 sql.(*tableState).release.func1 { m.isDraining() || }
lease.go:751 sql.(*tableState).release { if decRefcount(s) { }
lease.go:1153 sql.(*LeaseManager).Release { return t.release(lease, m) }
lease_test.go:133 sql_test.(*leaseTest).release { return t.node(nodeID).Release(lease) }
lease_test.go:147 sql_test.(*leaseTest).mustRelease { if err := t.release(nodeID, lease); err != nil { }
lease_test.go:427 sql_test.TestLeaseManagerDrain { t.mustRelease(1, l1, nil) }

in another goroutine: happened before
lease.go:1165 sql.(*LeaseManager).SetDraining { m.mu.Lock() } <<<<<
lease_test.go:434 sql_test.TestLeaseManagerDrain { t.nodes[1].SetDraining(true) }

happend after
lease.go:1172 sql.(*LeaseManager).SetDraining { t.mu.Lock() } <<<<<
lease_test.go:434 sql_test.TestLeaseManagerDrain { t.nodes[1].SetDraining(true) }


Other goroutines holding locks:
goroutine 2279390 lock 0xc42b457328
lease.go:1165 sql.(*LeaseManager).SetDraining { m.mu.Lock() } <<<<<
lease_test.go:434 sql_test.TestLeaseManagerDrain { t.nodes[1].SetDraining(true) }




ERROR: exit status 2

make: *** [stress] Error 1
1 runs completed, 1 failures, over 56s
Makefile:147: recipe for target 'stress' failed
```

### PR DESCRIPTION
**Pull request by asubiotto**
Fixes #12903 

cc @a-robinson

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12948)

### CODE REVIEW

File: pkg/sql/lease.go [hunk 1]
```diff
@@ -956,9 +957,10 @@ func makeTableNameCacheKey(dbID sqlbase.ID, tableName string) tableNameCacheKey
 // LeaseManager.mu > tableState.mu > tableNameCache.mu > LeaseState.mu
 type LeaseManager struct {
 	LeaseStore
-	mu       syncutil.Mutex
-	draining bool
-	tables   map[sqlbase.ID]*tableState
+	mu     syncutil.Mutex
```
Comment (tamird) [inline pkg/sql/lease.go]: while you're here, can you make this an anonymous struct so it's clear what `mu` is protecting?

**Review by tamird — APPROVED** [review]
This commit message is quite anemic. What is the lock ordering violation you're fixing?

File: pkg/sql/lease.go (hunk 1 above)
Comment (asubiotto) [inline pkg/sql/lease.go]: Done. Updated commit message too.

**Review by a-robinson — APPROVED** [review]
LGTM, thanks!

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/lease.go, line 1192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948#-KahJjmQA1M3Ga1vK2-7:-KahJjmQA1M3Ga1vK2-8:brvjx3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/998a35cf561bb01ae59fd2b87659061644073aba/pkg/sql/lease.go#L1192)):*
> ```Go
> 	if t == nil && create {
> 		t = &tableState{id: tableID, tableNameCache: &m.tableNames, stopper: m.stopper}
> 		m.mu.tables[tableID] = t
> ```

The locking still looks sketchy - we insert into `m.mu.tables` without checking the draining state here, so there's no guarantee that the loop in `SetDraining` will see everything. I guess it's no worse than before (since `isDraining` doesn't compose with any other critical sections), but it doesn't look right. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948)*

**Comment by asubiotto** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/lease.go, line 1192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948#-KahJjmQA1M3Ga1vK2-7:-KahMhNqIz-uKCPBWtmX:b-oc5vwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/998a35cf561bb01ae59fd2b87659061644073aba/pkg/sql/lease.go#L1192)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The locking still looks sketchy - we insert into `m.mu.tables` without checking the draining state here, so there's no guarantee that the loop in `SetDraining` will see everything. I guess it's no worse than before (since `isDraining` doesn't compose with any other critical sections), but it doesn't look right. 
</blockquote></details>

As discussed in person, it is only important to notice tables if there are leases to be released. Even if we don't notice it in the `SetDraining` loop, no acquisitions will be allowed to take place.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948)*

**Comment by bdarnell** [comment]
LGTM to fix the nightly test spam. 

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/lease.go, line 1192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948#-KahJjmQA1M3Ga1vK2-7:-KahRkqGn1OaVbBGf2uT:b6eh4hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/998a35cf561bb01ae59fd2b87659061644073aba/pkg/sql/lease.go#L1192)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

As discussed in person, it is only important to notice tables if there are leases to be released. Even if we don't notice it in the `SetDraining` loop, no acquisitions will be allowed to take place.
</blockquote></details>

Is that true? In `tableState.acquireNodeLease`, we check `isDraining` at the start of the function then unlock `t.mu`. It doesn't look like this is properly synchronized with the draining operation. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948#-:-KahSJEU0_csdqwtWQRl:b6tmp6e)*

**Comment by vivekmenezes** [comment]
LGTM

**Comment by asubiotto** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/lease.go, line 1192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948#-KahJjmQA1M3Ga1vK2-7:-KahV9PBAnJqGjworIqm:b-v01ydh) ([raw file](https://github.com/cockroachdb/cockroach/blob/998a35cf561bb01ae59fd2b87659061644073aba/pkg/sql/lease.go#L1192)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is that true? In `tableState.acquireNodeLease`, we check `isDraining` at the start of the function then unlock `t.mu`. It doesn't look like this is properly synchronized with the draining operation. 
</blockquote></details>

Sorry, you're right. Some acquisitions can take place when draining is set because `t.mu` is unlocked when acquiring leases. This happens when the draining check passes but the acquisition doesn't take place until after the corresponding `tableState`'s iteration in `SetDraining`. However, this lease will be deleted when released and can be said to be similar to holding a lease before draining so this behavior is manageable. I will probably refactor this in the future to make the lease code clearer and locking simpler.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12948)*

**Pull request merged**